### PR TITLE
all: introduce work.BlockChain interface

### PR DESCRIPTION
**Pull request by ehnuje**
## Proposed changes

- `work.BlockChain` interface has been introduced to test `cn.CN` and `sc.ServiceChain`

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by aidan-kwon** [comment]
@ehnuje Could you change the filename "work/mocks/blockchain_mock.go" to "work/mocks/blockchain_mock_test.go"?

**Comment by ehnuje** [comment]
@aidan-kwon Files under the `mocks` package should have the names without `_test` so that they can be seen outside the `mocks` package, don't they?

**Comment by KimKyungup** [comment]
Just curious. 
`cn.CN` and `sc.ServiceChain` need different methods of blockchain.
In this case, using same `work.BlockChain` interface is better than using their own each different interface of `blockchain`?

**Comment by ehnuje** [comment]
@KimKyungup Good point =) This is somewhat similar situation you asked in https://github.com/klaytn/klaytn/pull/230#discussion_r325481841. 

- If we want to separate the usages of `cn.CN` and `sc.ServiceChain`, `work.BlockChain` shouldn't be used in `work.Backend` as `cn.CN` and `sc.ServiceChain` need to implement it. 
- If so, we need to wrapping the calls of `Backend.BlockChain()`
- Also, as `work.worker` also has its own member variable `chain` received from `Backend`, we also need to remove the usages of `chain` from `work.worker` too. From this point, there needs to be huge changes to make it work (I mean, build)
- I also agree with your point using different interface for different usages, however, it needs quite a lot of work to do and makes `Backend` interface huge.

**Comment by aidan-kwon** [comment]
> @aidan-kwon Files under the `mocks` package should have the names without `_test` so that they can be seen outside the `mocks` package, don't they?

Yes for the coverage test.  
I put that comment since thought all mocking files have "*_mock_test.go" name for the consistency. However, your file name also looks good to me. Never mind my first comment.

File: blockchain/state_processor.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 // and uses the input parameters for its environment. It returns the receipt
 // for the transaction, gas used and an error if the transaction failed,
 // indicating the block was invalid.
-func ApplyTransaction(config *params.ChainConfig, bc *BlockChain, author *common.Address, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg *vm.Config) (*types.Receipt, uint64, error) {
+func (bc *BlockChain) ApplyTransaction(config *params.ChainConfig, author *common.Address, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg *vm.Config) (*types.Receipt, uint64, error) {
```
Comment (aidan-kwon) [inline blockchain/state_processor.go]: How about moving this function to "/blockchain/blockchain.go" file?

File: node/cn/protocol.go [hunk 2]
```diff
@@ -134,41 +132,6 @@ type TxPool interface {
 	SubscribeNewTxsEvent(chan<- blockchain.NewTxsEvent) event.Subscription
 }
 
-//go:generate mockgen -destination=node/cn/mocks/blockchain_mock.go -package=mocks github.com/klaytn/klaytn/node/cn BlockChain
```
Comment (aidan-kwon) [inline node/cn/protocol.go]: Will you leave "node/cn/mocks/blockchain_mock.go" file?

Comment (ehnuje) [reply node/cn/protocol.go]: My mistake. I removed it. Thanks!

File: blockchain/state_processor.go (hunk 1 above)
Comment (ehnuje) [reply blockchain/state_processor.go]: That sounds better. Updated!

**Comment by ehnuje** [comment]
@KimKyungup You're right. I force pushed since this needs rebasing.

File: tests/pregenerated_data_util_test.go [hunk 3]
```diff
@@ -378,7 +378,7 @@ func NewBCDataForPreGeneratedTest(testDataDir string, tc *preGeneratedTC) (*BCDa
 	engine := istanbulBackend.New(genesisAddr, istanbul.DefaultConfig, validatorPrivKeys[0], chainDB, gov, node.CONSENSUSNODE)
 
 	////////////////////////////////////////////////////////////////////////////////
-	// Make a BlockChain
+	// Make a blockChain
```
Comment (jeonghonah) [inline tests/pregenerated_data_util_test.go]: Why is this changed?

Comment (ehnuje) [reply tests/pregenerated_data_util_test.go]: Looks like Goland's refactoring feature has changed it automatically...
